File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
#include<string.h>
using namespace std;
bool visited[104][104]; // 방문여부
int dist[104][104]; // 거리
int dx[4] = {1,-1,0,0}; // 방향
int dy[4] = {0,0,1,-1}; // 방향
queue<pair<int,int>> q;

int solution(vector<vector<int> > maps)
{   
    int answer = 0;
    int n = maps.size();
    int m = maps[0].size();
    memset(dist,-1,sizeof(dist)); // -1로 초기화
    q.push({0,0});
    visited[0][0] = 1;
    dist[0][0] = 1;
    
    while(!q.empty()){
        auto cur = q.front(); q.pop();
        // 종료조건
        if(cur.first == n -1 && cur.second == m -1){
            return dist[n-1][m-1];
        }
        for(int dir = 0; dir < 4; dir++){
            int nx = cur.second + dx[dir];
            int ny = cur.first + dy[dir];
            if(nx < 0 || ny < 0 || nx >= m || ny >= n) continue;
            if(visited[ny][nx] || maps[ny][nx] == 0 || dist[ny][nx] != -1) continue;
            q.push({ny,nx});
            visited[ny][nx];
            dist[ny][nx] = dist[cur.first][cur.second] + 1 ;
        }
    }
    if(dist[n-1][m-1] == -1) return -1;
    else return dist[n-1][m-1];
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
# 최단거리, 동일 비용 = bfs 로 구현
from collections import deque

def solution(maps):
    queue = deque([(0,0)])
    n,m = len(maps), len(maps[0])
    
    while queue:
        x, y = queue.popleft()
        for dx, dy in ((1,0),(-1,0),(0,1),(0,-1)):
            nx, ny = x + dx, y + dy
        
            if 0 <= nx < n and 0 <= ny < m and maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                queue.append((nx,ny))
                
    return maps[n-1][m-1] if maps[n-1][m-1] > 1 else -1
